### PR TITLE
feat(php): Design multipart/form-data requests

### PR DESCRIPTION
**Pull request by amckinney**
This outlines an approach to support `multipart/form-data` requests with the following new types:

* `Core\MultipartApiRequest`
* `Core\MultipartFormData`
* `Core\MultipartFormDataPart`
* `Utils\File` (user-facing)

References:
  * Guzzle multipart requests ([link](https://docs.guzzlephp.org/en/latest/request-options.html#multipart))

## UX

Users are presented with a generated request type that specifies the `Utils\File` type alongside any other in-lined request parameters.

```php
<?php

namespace Seed\Service\Requests;

use Acme\Utils\File;

class FileUploadRequest
{
    /**
     * @var File $file
     */
    public File $file;

    /**
     * @var User $user
     */
    public User $user;

    /**
     * @param array{
     *   file: File,
     *   user: User,
     * } $values
     */
    public function __construct(
        array $values,
    ) {
        $this->file = $values['file'];
        $this->user = $values['user'];
    }
}
```

With this, users can create the `File` type by using the primary constructor or any of the other static constructor helper methods, and call the method like so:

```php
$client->files->upload(
  new FileUploadRequest(
    [
      'file' => File::createFromFilepath('./path/to/example.png'),
      'user' => new User(['name' => 'john'])
    ]
  )
);
```

## Implementation

Behind the scenes, the new `Utils\File` type includes a mapper method for the `Core\MultipartFormDataPart`:

```php
/**
 * Maps this File into a multipart form data part.
 *
 * @param string $name The name of the mutlipart form data part.
 * @param ?string $contentType Overrides the Content-Type associated with the file, if any.
 * @return MultipartFormDataPart
 */
public function toMultipartFormDataPart(string $name, ?string $contentType = null): MultipartFormDataPart
{
    $contentType ??= $this->contentType;
    $headers = $contentType != null
        ? ['Content-Type' => $contentType]
        : null;

    return new MultipartFormDataPart(
        name: $name,
        contents: $this->stream,
        filename: $this->filename,
        headers: $headers,
    );
}
```

The `MultipartFormDataPart` is a class used to represent individual multipart form data elements in the Guzzle API ([docs](https://docs.guzzlephp.org/en/latest/request-options.html#multipart)). It includes a `toArray` method that maps into this form that looks like the following:

```php
/**
 * Converts the multipart form data part into an array suitable
 * for Guzzle's multipart form data.
 *
 * @return array{
 *     name: string,
 *     contents: StreamInterface,
 *     filename?: string,
 *     headers?: array<string, string>
 * }
 */
public function toArray(): array
{
    $formData = [
        'name' => $this->name,
        'contents' => $this->contents,
    ];

    if ($this->filename != null) {
        $formData['filename'] = $this->filename;
    }

    if ($this->headers != null) {
        $formData['headers'] = $this->headers;
    }

    return $formData;
}
```

The `MultipartFormData` is simply a collection of `MultipartFormDataPart` elements with additional methods for constructing and adding these parts. It is mapped to Guzzle's representation with another `toArray` method that looks like the following:

```php
/**
 * Converts the multipart form data into an array suitable
 * for Guzzle's multipart form data.
 *
 * @return array<array{
 *     name: string,
 *     contents: StreamInterface,
 *     filename?: string,
 *     headers?: array<string, string>
 * }>
 */
public function toArray(): array
{
    return array_map(fn($part) => $part->toArray(), $this->parts);
}
```

Finally, the endpoint is implemented so that we compose a `MultipartFormData` element, and individually add parts based on their wire name (e.g. `'number'` for the `number` property):

```php
/**
 * @param UploadFileRequest $request
 * @param ?array{
 *   baseUrl?: string,
 * } $options
 * @throws AcmeException
 * @throws AcmeApiException
 */
public function uploadFIle(UploadFileRequest $request, ?array $options = null): void
{
    try {
        $body = new MultipartFormData();
        $body->addPart($request->file->toMultipartFormDataPart('file'));
        $body->add('user', $request->user->toJson());
    } catch (Exception $e) {
        throw new AcmeException(message: $e->getMessage(), previous: $e);
    }

    try {
        $response = $this->client->sendRequest(
            new JsonApiRequest(
                baseUrl: $this->options['baseUrl'] ?? $this->client->options['baseUrl'] ?? '',
                path: "/file/upload",
                method: HttpMethod::POST,
                body: $body,
            ),
        );
        $statusCode = $response->getStatusCode();
        if ($statusCode >= 200 && $statusCode < 400) {
            return;
        }
    } catch (ClientExceptionInterface $e) {
        throw new AcmeException(message: $e->getMessage(), previous: $e);
    }
    throw new AcmeApiException(
        message: 'API request failed',
        statusCode: $statusCode,
        body: $response->getBody()->getContents(),
    );
}
```

### CODE REVIEW

File: seed/php-sdk/file-upload/src/Utils/File.php [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+    private ?string $contentType;
+
+    /**
+     * @param StreamInterface $stream
+     * @param ?string $filename
+     * @param ?string $contentType
+     */
+    public function __construct(
+        StreamInterface $stream,
+        ?string $filename,
```
Comment (dcb6) [inline seed/php-sdk/file-upload/src/Utils/File.php:31]: nit:
```suggestion
        ?string $filename = null,
```

File: seed/php-sdk/file-upload/src/Utils/File.php [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+     * @param ?string $contentType
+     * @return File
+     * @throws Exception
+     */
+    public static function createFromFilepath(
+        string $filepath,
+        ?string $filename = null,
+        ?string $contentType = null,
+    ): File {
+        $stream = Utils::streamFor(fopen($filepath, 'r'));
```
Comment (dcb6) [inline seed/php-sdk/file-upload/src/Utils/File.php:53]: looks like `fopen` returns `resource|false`
maybe should split out:
```
    $resource = fopen($filepath, 'r');
    if ($resource === false) {
        throw new Exception("Unable to open the file: $filepath");
    }
    $stream = Utils::streamFor($resource);
```

**Review by dcb6 — APPROVED** [review]
left a couple small comments! i don't have much experience with file upload to pull from, but all of this looks great from where I stand 😊

**Comment by amckinney** [comment]
This design was implemented in https://github.com/fern-api/fern/pull/5030 and released in version `0.2.0` [here](https://github.com/fern-api/fern/pull/5051).

**Pull request closed without merging**